### PR TITLE
fix: use ceiling division for veto threshold to prevent rounding to zero

### LINKED ISSUE #24: Veto threshold floor division can round to zero, unconditionally defeating proposals
# Security Disclosure: Veto Threshold Floor Division Can Cancel Proposals Unconditionally

**Repository:** reserve-protocol/reserve-governor
**Severity:** Low
**Status:** Unfixed as of 2026-03-28

## Summary

The veto threshold calculation in `ReserveOptimisticGovernor` uses Solidity's floor division, which can round the token-denominated threshold to zero. When this happens, the contract unconditionally cancels the proposal regardless of veto votes — effectively making any proposal vetoed by default.

## Affected Code

**File:** `contracts/governance/ReserveOptimisticGovernor.sol`, lines 247-249

```solidity
uint256 vetoThresholdTok = (_vetoThreshold * token().getPastTotalSupply(snapshot)) / 1e18;

if (vetoThresholdTok == 0) {
    return ProposalState.Defeated;  // Proposal unconditionally canceled
}
```

`_vetoThreshold` is a D18 fraction (constrained to `0 < vetoThreshold <= 1e18` by `_setOptimisticParams`). The floor division rounds down, so when `_vetoThreshold * totalSupply < 1e18`, the result is zero and the proposal is automatically defeated.

## Impact

**Practically limited:** For `vetoThresholdTok` to round to zero, you'd need either:
- An extremely small `_vetoThreshold` (e.g., `1` out of `1e18`), AND/OR
- A very small `totalSupply` (e.g., less than `1e18` wei of the governance token)

In any real deployment with reasonable parameters, this edge case is unreachable. However, the floor division does make the effective veto threshold slightly lower than the mathematically exact value, marginally easing vetoes in all cases.

## Suggested Fix

Use ceiling division to ensure the threshold is never rounded down to zero:

```diff
--- a/contracts/governance/ReserveOptimisticGovernor.sol
+++ b/contracts/governance/ReserveOptimisticGovernor.sol
@@ -244,7 +244,9 @@
         {
             uint256 snapshot = proposalSnapshot(proposalId);

-            uint256 vetoThresholdTok = (_vetoThreshold * token().getPastTotalSupply(snapshot)) / 1e18;
+            // Use ceiling division to prevent rounding to zero
+            uint256 product = _vetoThreshold * token().getPastTotalSupply(snapshot);
+            uint256 vetoThresholdTok = (product + 1e18 - 1) / 1e18;

             if (vetoThresholdTok == 0) {
                 return ProposalState.Defeated;
```

This ensures a nonzero `_vetoThreshold` always produces a nonzero token threshold (minimum 1 wei), and the effective threshold is never lower than intended.

### PR DESCRIPTION
**Pull request by omaidf**
## Summary

- Replaces floor division with ceiling division in veto threshold calculation
- Prevents `vetoThresholdTok` from rounding to zero and unconditionally canceling proposals

## Problem

At line 247 of `ReserveOptimisticGovernor.sol`:

```solidity
uint256 vetoThresholdTok = (_vetoThreshold * token().getPastTotalSupply(snapshot)) / 1e18;
```

Floor division can produce 0 when `_vetoThreshold * totalSupply < 1e18`. Line 249 then cancels the proposal unconditionally when `vetoThresholdTok == 0`. With ceiling division, a nonzero `_vetoThreshold` always produces at least 1 wei of token threshold.

Fixes #24

## Test plan

- [ ] Proposals with normal parameters behave identically (ceiling vs floor produces same result for large totalSupply)
- [ ] Edge case: very small `_vetoThreshold` with small `totalSupply` no longer auto-cancels

### CODE REVIEW

**Comment by tbrent** [comment]
This rounding direction is intended for defensiveness:
- Higher threshold → harder to veto → worst case permanent loss
- Lower threshold → easier to veto → worst case optimistic DOS, pessimistic backup

Thanks for opening the PR, though!
